### ISSUE TITLE
Should userdata folder be part of version control?

### ISSUE THREAD

**Issue by tnguyen14** [coreos/kube-aws#236]
I am wondering if the `userdata` folder should be part of git, or whether it should be gitignored.

@robszumski says that it can be used as-is https://twitter.com/robszumski/status/818902908448014337, but I'm not sure what that means for version control.

The problem I see with having it in git is that if someone else wants to do `kube-aws up`, it will throw an error if the `userdata` folder is there. It seems to me that both `credentials` and `userdata` folders need to be automatically generated.

I guess this begs the question of the purpose of keeping things in git. My original intention is to keep the `cluster.yml` file, so that someone else could spin up the same Kubernetes cluster. For that same purpose, would the other files be useful? (`kubeconfig`, `stack-template.json`, `[cluster].stack-template.json`).

I also intend to store the different kubernetes deployment and service manifest yaml files in the same git repository.

If the purpose of the repo was so that other people could collaborate on the same cluster (and not necessary spinning up a new one), what files/ folders would make sense to be included?

**Comment by redbaron** [comment]
serdata is coupled with content of stack-template, for example `AWS::CloudFormation::Init` resource in `stack-template.json` is used by `cfn-init` scripts in userdata. So they got to be in sync and kept together

**Comment by robszumski** [comment]
@tnguyen14 From your original tweet I thought you were worried about editing it. As @redbaron stated, you will need both the user-data and the stack template in order to `up`. It's fine to keep it under version control as all of the sensitive bits are templated in as encrypted values.

**Comment by redbaron** [comment]
@tnguyen14 maybe I lack context, but why deleting userdata is needed to create a new cluster?

**Comment by tnguyen14** [comment]
Oh, hmm, let me clarify.

When someone else wants to create a new cluster, they would need to run `kube-aws render` to create their own `credentials` folder, is that correct? If so, this command will return an error if the `userdata` folder already exists, even though the `credentials` folder is still created.

**Comment by redbaron** [comment]
credentials are not included into `userdata` folder, files there are templates  which are dynamically templated when you run `up/update`.

So if you want have a new set of credentials you can have it, just run `kube-aws render credentials && rm credentials/*.enc`

**Comment by tnguyen14** [comment]
yeah I know `credentials` are not in `userdata`. I just see the behavior that running `kube-aws render` will error if either of `credentials` or `userdata` exists.

so will `kube-aws render credentials` not check for `userdata`? and why `rm credentials/*.enc`?

**Comment by redbaron** [comment]
which version do you use? I can run `[[ -d credentials ]] && kube-aws render stack` as much as I want and there is no error.

**Comment by tnguyen14** [comment]
```shell
:; kube-aws version
kube-aws version 0.8.3
```

seems to be the latest?

**Comment by redbaron** [comment]
use latest RC from releases page

**Comment by tnguyen14** [comment]
Does `kube-aws render credentials` create SSL certificates by default?

**Comment by redbaron** [comment]
you need to pass either `--generate-ca` or point it to CA cert and key

**Comment by mumoshu** [comment]
Closing this as original question is answered(thanks to @redbaron :bow:)

**Comment by tnguyen14** [comment]
@redbaron if the cluster is already up and running with a CA, each new developer who wants to work on it will need to run `kube-aws render credentials --ca-cert-path= --ca-key-path=` to generate their own `credentials` folder, is that correct?

**Comment by redbaron** [comment]
no, it won't work.kube-aws shies away from 3 real life cases, leaving it up to you to , solve it

1. etcd management
2. certs management
3. rolling upgrades

Regarding certs management, if multiple devs tried to work on same cluster, but use their own credentials, this essentially be a cert rotation, which wont work out of the box because API servers will be started effectively with new  `--service-account-key-file` argument (which by default points to apiserver-key.pem), but existing serviceaccount tokens were generated previous version of key, therefore all pods wont be able to speak to newly started APIservers as they wont authenticate their requests, because keys don't match.

Long story short, once cluster is up, use same certs. I think it is fairly safe to commit .enc files as they are encrypted with KMS key.

**Comment by tnguyen14** [comment]
@redbaron so in `credentials` folder, I see different kinds of certs (`ca`, `api-server`, `worker` and `admin`). When you say "certs", do you refer to all of them, or just the `ca` certs?

**Comment by redbaron** [comment]
well ,you asked about `--ca-cert-path= --ca-key-path=` args, so I assumed that CA cert and key will be shared between devs, so I was refering to `api-server` only, it got to be shared as well.

If you intend to share nothing and let every dev to generate full stack of certs, then nothing will work, as `kube-aws` leaves Etcd servers untouched after first up, therefore newly generated certs will not be recognised by Etcd servers continue to use CA cert from a dev who spun up the cluster first time.

**Comment by tnguyen14** [comment]
@redbaron thanks. for devs who are not the first time creator, if they use the original (shared) `ca` and `api-server` certs, would they still need to run `kube-aws render credentials`?

**Comment by tnguyen14** [comment]
Also, how should the `.enc` files be decrypted by other devs?
